File: python/second/min_stack.py
class MinStack(object):

    def __init__(self):
        self.stack = []
        

    def push(self, val):
        """
        :type val: int
        :rtype: None
        """
        if len(self.stack) == 0:
            self.stack.append((val, val))
        else:
            curMin = self.stack[-1][1]
            self.stack.append((val, min(val, curMin)))
        

    def pop(self):
        """
        :rtype: None
        """
        if len(self.stack) == 0:
            raise RuntimeError()
        self.stack.pop()


    def top(self):
        """
        :rtype: int
        """
        if len(self.stack) == 0:
            raise RuntimeError()
        return self.stack[-1][0]
        

    def getMin(self):
        """
        :rtype: int
        """
        if len(self.stack) == 0:
            raise RuntimeError()
        return self.stack[-1][1]
        

File: src/main/java/october_2025/MinStack.java
package october_2025;

import java.util.ArrayDeque;

public class MinStack {

    // O(1) time | O(n) space
    ArrayDeque<int[]> stack;

    public MinStack() {
        stack = new ArrayDeque<>();
    }

    public void push(int val) {
        int[] newElement = new int[2];
        if (stack.isEmpty()) {
            newElement[1] = val;
        } else {
            int top = stack.getLast()[1];
            newElement[1] = Math.min(top, val);
        }
        newElement[0] = val;
        stack.addLast(newElement);
    }

    public void pop() {
        if (stack.isEmpty()) {
            throw new RuntimeException("Stack is Empty!");
        }
        stack.removeLast();
    }

    public int top() {
        if (stack.isEmpty()) {
            throw new RuntimeException("Stack is Empty!");
        }
        return stack.getLast()[0];

    }

    public int getMin() {
        if (stack.isEmpty()) {
            throw new RuntimeException("Stack is Empty!");
        }
        return stack.getLast()[1];
    }

}
